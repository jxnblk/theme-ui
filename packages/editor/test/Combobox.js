import React from 'react'
import renderer from 'react-test-renderer'
import {
  render,
  fireEvent,
  act,
  waitForElement,
  waitForDomChange,
  cleanup
} from '@testing-library/react'
import Combobox from '../src/Combobox'

afterEach(cleanup)

test('renders', () => {
  const json = renderer.create(
    <Combobox />
  ).toJSON()
  expect(json).toMatchSnapshot()
})

test('clicking chevron button shows menu', async () => {
  const tree = render(
    <Combobox
      label='Beep'
      name='beep'
      options={[
        'beep',
        'boop',
      ]}
    />
  )
  const button = await tree.getByRole('button')
  fireEvent.click(button)
  const options = await waitForElement(() => tree.findAllByRole('option'))
  expect(options.length).toBe(2)
})

test('clicking item updates value', async () => {
  const onChange = jest.fn()
  const tree = render(
    <Combobox
      label='Beep'
      name='beep'
      options={[
        'beep',
        'boop',
      ]}
      onChange={onChange}
    />
  )
  const button = await tree.getByRole('button')
  fireEvent.click(button)
  const [ option ] = await waitForElement(() => tree.findAllByRole('option'))
  fireEvent.click(option)
  expect(onChange).toHaveBeenCalledWith('beep')
})

test('down arrow moves selection down', async () => {
  const onChange = jest.fn()
  const tree = render(
    <Combobox
      label='Beep'
      name='beep'
      options={[
        'beep',
        'boop',
      ]}
      onChange={onChange}
    />
  )
  const button = await tree.getByRole('button')
  fireEvent.click(button)
  const input = await tree.getByLabelText('Beep')
  fireEvent.keyDown(input, {
    key: 'ArrowDown',
    keyCode: 40,
  })
  fireEvent.keyDown(input, {
    key: 'ArrowDown',
    keyCode: 40,
  })
  fireEvent.keyDown(input, {
    key: 'Enter',
    keyCode: 13,
  })
  expect(onChange).toHaveBeenCalledWith('boop')
})

test('up arrow moves selection up', async () => {
  const onChange = jest.fn()
  const tree = render(
    <Combobox
      label='Beep'
      name='beep'
      options={[
        'beep',
        'boop',
      ]}
      onChange={onChange}
    />
  )
  const button = await tree.getByRole('button')
  fireEvent.click(button)
  const input = await tree.getByLabelText('Beep')
  fireEvent.keyDown(input, {
    key: 'ArrowDown',
    keyCode: 40,
  })
  fireEvent.keyDown(input, {
    key: 'ArrowDown',
    keyCode: 40,
  })
  fireEvent.keyDown(input, {
    key: 'ArrowUp',
    keyCode: 38,
  })
  fireEvent.keyDown(input, {
    key: 'Enter',
    keyCode: 13,
  })
  expect(onChange).toHaveBeenCalledWith('beep')
})

test('escape key closes listbox', async () => {
  const tree = render(
    <Combobox
      label='Beep'
      name='beep'
      options={[
        'beep',
        'boop',
      ]}
    />
  )
  const button = await tree.getByRole('button')
  const input = await tree.getByLabelText('Beep')
  const listbox = await tree.getByRole('listbox', {
    hidden: true,
  })
  fireEvent.click(button)
  fireEvent.keyDown(input, {
    key: 'Escape',
    keyCode: 27,
  })
  expect(listbox.style.visibility).toBe('hidden')
})

test('down arrow key opens listbox', async () => {
  const tree = render(
    <Combobox
      label='Beep'
      name='beep'
      options={[
        'beep',
        'boop',
      ]}
    />
  )
  const input = await tree.getByLabelText('Beep')
  const listbox = await tree.getByRole('listbox', {
    hidden: true,
  })
  fireEvent.keyDown(input, {
    key: 'ArrowDown',
    keyCode: 40,
  })
  expect(listbox.style.visibility).toBe('visible')
})

test('up arrow key opens listbox', async () => {
  const tree = render(
    <Combobox
      label='Beep'
      name='beep'
      options={[
        'beep',
        'boop',
      ]}
    />
  )
  const input = await tree.getByLabelText('Beep')
  const listbox = await tree.getByRole('listbox', {
    hidden: true,
  })
  fireEvent.keyDown(input, {
    key: 'ArrowUp',
    keyCode: 38,
  })
  expect(listbox.style.visibility).toBe('visible')
})

test('ignores keypresses with meta key', async () => {
  const tree = render(
    <Combobox
      label='Beep'
      name='beep'
      options={[
        'beep',
        'boop',
      ]}
    />
  )
  const input = await tree.getByLabelText('Beep')
  const listbox = await tree.getByRole('listbox', {
    hidden: true,
  })
  fireEvent.keyDown(input, {
    key: 'ArrowDown',
    keyCode: 40,
    metaKey: true
  })
  expect(listbox.style.visibility).toBe('hidden')
})

test('ignores return key when closed', async () => {
  const tree = render(
    <Combobox
      label='Beep'
      name='beep'
      options={[
        'beep',
        'boop',
      ]}
    />
  )
  const input = await tree.getByLabelText('Beep')
  const listbox = await tree.getByRole('listbox', {
    hidden: true,
  })
  fireEvent.keyDown(input, {
    key: 'Return',
    keyCode: 13,
  })
  expect(listbox.style.visibility).toBe('hidden')
})

// todo
test.skip('blur closes listbox', async () => {
  global.requestAnimationFrame = jest.fn(fn => setTimeout(fn, 1))
  const tree = render(
    <Combobox
      label='Beep'
      name='beep'
      options={[
        'beep',
        'boop',
      ]}
    />
  )
  const button = await tree.getByRole('button')
  const input = await tree.getByLabelText('Beep')
  fireEvent.click(button)
  fireEvent.blur(input, {})
  const listbox = await tree.getByRole('listbox', {
    hidden: true,
  })
  expect(global.requestAnimationFrame).toHaveBeenCalled()
  // not sure how to use testing-library here
  await waitForDomChange({ container: tree })
  expect(listbox.style.visibility).toBe('hidden')
})
