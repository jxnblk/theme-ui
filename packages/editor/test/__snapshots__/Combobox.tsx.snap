// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
.emotion-0 {
  z-index: 2;
}

.emotion-1 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-3 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  display: block;
  width: 100%;
  padding: 8px;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  font-size: inherit;
  line-height: inherit;
  border: 1px solid;
  border-radius: 4px;
  color: inherit;
  background-color: transparent;
}

.emotion-3:autofill,
.emotion-3:autofill:hover,
.emotion-3:autofill:focus {
  box-shadow: inset 0 0 0 1000px var(--theme-ui-input-autofill-bg);
  font-size: inherit;
}

.emotion-3:autofill:first-line,
.emotion-3:autofill:hover:first-line,
.emotion-3:autofill:focus:first-line {
  font-size: 1rem;
}

.emotion-3:-webkit-autofill,
.emotion-3:-webkit-autofill:hover,
.emotion-3:-webkit-autofill:focus {
  box-shadow: inset 0 0 0 1000px var(--theme-ui-input-autofill-bg);
  font-size: inherit;
}

.emotion-3:-webkit-autofill:first-line,
.emotion-3:-webkit-autofill:hover:first-line,
.emotion-3:-webkit-autofill:focus:first-line {
  font-size: 1rem;
}

.emotion-4 {
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  width: 32px;
  height: 32px;
  color: inherit;
  background-color: transparent;
  border: 0;
  margin-left: -32px;
}

.emotion-5 {
  pointer-events: none;
}

.emotion-6 {
  left: 0;
  right: 0;
  top: 100%;
  list-style: none;
  padding: 0;
  margin: 0;
  max-height: 512px;
  overflow-y: auto;
  background-color: white;
  color: black;
  border: 1px solid;
  border-top: 0;
  border-color: lightgray;
}

<div
  className="emotion-0"
  style={
    Object {
      "position": "static",
    }
  }
>
  <label
    className="emotion-1"
    htmlFor="beep"
  >
    beep
  </label>
  <div
    className="emotion-2"
  >
    <input
      aria-activedescendant="beep-1"
      aria-autocomplete="none"
      aria-expanded={false}
      aria-haspopup="true"
      aria-owns="beep-popup"
      className="emotion-3"
      id="beep"
      name="beep"
      onBlur={[Function]}
      onChange={[Function]}
      onKeyDown={[Function]}
      role="combobox"
      type="text"
      value="beep"
    />
    <button
      aria-label="Open"
      className="emotion-4"
      onClick={[Function]}
      tabIndex={-1}
    >
      <svg
        className="emotion-5"
        height="12"
        viewBox="0 0 16 16"
        width="12"
      >
        <path
          d="M14 6 L8 12 L2 6"
          fill="none"
          stroke="currentcolor"
          strokeWidth="2"
        />
      </svg>
    </button>
  </div>
  <ul
    aria-label="beep"
    className="emotion-6"
    id="beep-popup"
    role="listbox"
    style={
      Object {
        "position": "static",
        "visibility": "hidden",
      }
    }
    tabIndex={-1}
  />
</div>
`;
